File: apps/mobile/plugin/native/netease-player/android/src/main/java/com/mugame/mobile/plugins/NeteasePlayerPlugin.java
package com.mugame.mobile.plugins;

import android.content.ActivityNotFoundException;
import android.content.Intent;
import android.net.Uri;
import android.provider.Settings;
import android.util.Log;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

@CapacitorPlugin(name = "NeteasePlayer")
public class NeteasePlayerPlugin extends Plugin {
    private static final String TAG = "MuGameNeteasePlayer";
    private static final String NETEASE_PACKAGE = "com.netease.cloudmusic";
    private static final String PLAYLIST_URL_PREFIX = "orpheus://playlist/";
    private static final String SONG_URL_PREFIX = "orpheus://song/";
    private static final String AUTOPLAY_SUFFIX = "/?autoplay=1";

    private String currentTrackId;
    private String preparedUrl;
    private boolean preparedPlaylist;
    private boolean externalOpened;
    private String lastError;
    private NeteaseFloatingSourceWindow floatingSourceWindow;
    private final NeteasePlaybackMonitorService.MetadataListener metadataListener =
        snapshot -> notifyListeners("neteasePlaybackMetadataChanged", snapshot.toJson());

    @Override
    public void load() {
        floatingSourceWindow = new NeteaseFloatingSourceWindow(getContext());
        NeteasePlaybackMonitorService.addListener(metadataListener);
    }

    @Override
    protected void handleOnDestroy() {
        if (floatingSourceWindow != null) {
            floatingSourceWindow.hide();
        }
        NeteasePlaybackMonitorService.removeListener(metadataListener);
        super.handleOnDestroy();
    }

    @PluginMethod
    public void initialize(PluginCall call) {
        Log.i(TAG, "android external player bridge initialized");
        call.resolve();
    }

    @PluginMethod
    public void isPlaybackMonitorEnabled(PluginCall call) {
        JSObject result = new JSObject();
        result.put("enabled", NeteasePlaybackMonitorService.isEnabled(getContext()));
        call.resolve(result);
    }

    @PluginMethod
    public void configureSourceReveal(PluginCall call) {
        if (floatingSourceWindow != null) {
            floatingSourceWindow.updateTracks(call.getArray("tracks"));
        }
        call.resolve();
    }

    @PluginMethod
    public void isFloatingWindowEnabled(PluginCall call) {
        JSObject result = new JSObject();
        result.put("enabled", floatingSourceWindow != null && floatingSourceWindow.canShow());
        call.resolve(result);
    }

    @PluginMethod
    public void openFloatingWindowSettings(PluginCall call) {
        Intent intent = new Intent(
            Settings.ACTION_MANAGE_OVERLAY_PERMISSION,
            Uri.parse("package:" + getContext().getPackageName())
        );
        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        getContext().startActivity(intent);
        call.resolve();
    }

    @PluginMethod
    public void openPlaybackMonitorSettings(PluginCall call) {
        Intent intent = new Intent(Settings.ACTION_NOTIFICATION_LISTENER_SETTINGS);
        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        getContext().startActivity(intent);
        call.resolve();
    }

    @PluginMethod
    public void isPlaylistAutoplayEnabled(PluginCall call) {
        JSObject result = new JSObject();
        result.put("enabled", NeteasePlaylistAutoplayService.isEnabled(getContext()));
        call.resolve(result);
    }

    @PluginMethod
    public void openPlaylistAutoplaySettings(PluginCall call) {
        Intent intent = new Intent(Settings.ACTION_ACCESSIBILITY_SETTINGS);
        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        getContext().startActivity(intent);
        call.resolve();
    }

    @PluginMethod
    public void getCurrentPlaybackMetadata(PluginCall call) {
        call.resolve(NeteasePlaybackMonitorService.current(getContext()).toJson());
    }

    @PluginMethod
    public void ensureLoggedIn(PluginCall call) {
        Log.i(TAG, "external NetEase app owns playback login state");
        call.resolve();
    }

    @PluginMethod
    public void loadTrack(PluginCall call) {
        String songId = call.getString("netease_song_id");
        if (songId == null || songId.trim().isEmpty()) {
            call.reject("netease_song_id is required.", "invalid_track_id");
            return;
        }
        currentTrackId = songId.trim();
        preparedPlaylist = false;
        prepareExternalUrl(SONG_URL_PREFIX + currentTrackId + AUTOPLAY_SUFFIX);
        Log.i(TAG, "external NetEase song prepared id=" + currentTrackId);
        call.resolve();
    }

    @PluginMethod
    public void loadPlaylist(PluginCall call) {
        String playlistId = call.getString("netease_playlist_id");
        if (playlistId == null || playlistId.trim().isEmpty()) {
            call.reject("netease_playlist_id is required.", "invalid_playlist_id");
            return;
        }
        currentTrackId = null;
        preparedPlaylist = true;
        prepareExternalUrl(PLAYLIST_URL_PREFIX + playlistId.trim() + AUTOPLAY_SUFFIX);
        Log.i(TAG, "external NetEase playlist prepared id=" + playlistId.trim());
        call.resolve();
    }

    @PluginMethod
    public void play(PluginCall call) {
        if (preparedUrl == null) {
            call.reject("No NetEase playlist or song is prepared.", "player_not_initialized");
            return;
        }
        openPreparedUrl(call);
    }

    @PluginMethod
    public void pause(PluginCall call) {
        call.reject(
            "Transport controls are not available in external NetEase app mode.",
            "player_action_unsupported"
        );
    }

    @PluginMethod
    public void next(PluginCall call) {
        call.reject(
            "Transport controls are not available in external NetEase app mode.",
            "player_action_unsupported"
        );
    }

    @PluginMethod
    public void previous(PluginCall call) {
        call.reject(
            "Transport controls are not available in external NetEase app mode.",
            "player_action_unsupported"
        );
    }

    @PluginMethod
    public void seek(PluginCall call) {
        call.reject(
            "Seek is not available when playback is handled by NetEase app.",
            "player_action_unsupported"
        );
    }

    @PluginMethod
    public void getPlaybackState(PluginCall call) {
        if (externalOpened) {
            call.resolve(playbackState("playing"));
            return;
        }
        if (preparedUrl != null) {
            call.resolve(playbackState("paused"));
            return;
        }
        call.resolve(playbackState("idle"));
    }

    @PluginMethod
    public void destroy(PluginCall call) {
        currentTrackId = null;
        preparedUrl = null;
        preparedPlaylist = false;
        externalOpened = false;
        lastError = null;
        if (floatingSourceWindow != null) {
            floatingSourceWindow.hide();
        }
        call.resolve();
    }

    private void prepareExternalUrl(String url) {
        preparedUrl = url;
        externalOpened = false;
        lastError = null;
    }

    private void openPreparedUrl(PluginCall call) {
        getActivity().runOnUiThread(() -> {
            Intent appIntent = new Intent(Intent.ACTION_VIEW, Uri.parse(preparedUrl));
            appIntent.setPackage(NETEASE_PACKAGE);
            appIntent.addCategory(Intent.CATEGORY_BROWSABLE);
            appIntent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
            if (tryOpen(appIntent)) {
                externalOpened = true;
                lastError = null;
                Log.i(TAG, "external NetEase app opened");
                if (preparedPlaylist) {
                    NeteasePlaylistAutoplayService.requestPlaylistAutoplay();
                    if (floatingSourceWindow != null) {
                        floatingSourceWindow.show();
                    }
                }
                call.resolve();
                return;
            }

            Intent launchIntent = getContext().getPackageManager().getLaunchIntentForPackage(
                NETEASE_PACKAGE
            );
            if (launchIntent != null && tryOpen(launchIntent)) {
                externalOpened = true;
                lastError = "netease_deep_link_failed";
                Log.w(TAG, "NetEase app opened without playlist deep link");
                call.reject(
                    "NetEase app opened, but the playlist page could not be opened.",
                    "netease_deep_link_failed"
                );
                return;
            }

            lastError = "netease_app_open_failed";
            Log.w(TAG, lastError);
            call.reject("NetEase app could not be opened.", "netease_app_open_failed");
        });
    }

    private boolean tryOpen(Intent intent) {
        try {
            getContext().startActivity(intent);
            return true;
        } catch (ActivityNotFoundException error) {
            return false;
        }
    }

    private JSObject playbackState(String state) {
        JSObject result = new JSObject();
        result.put("state", state);
        result.put("currentTimeMs", 0);
        result.put("durationMs", 0);
        result.put("currentTrackId", currentTrackId);
        result.put("lastError", lastError);
        return result;
    }
}

File: apps/mobile/plugin/native/netease-player/ios/NeteasePlayerPlugin.swift
import Capacitor
import UIKit

@objc(NeteasePlayerPlugin)
public class NeteasePlayerPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "NeteasePlayerPlugin"
    public let jsName = "NeteasePlayer"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "initialize", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "ensureLoggedIn", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "configureSourceReveal", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "isFloatingWindowEnabled", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "openFloatingWindowSettings", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "isPlaybackMonitorEnabled", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "openPlaybackMonitorSettings", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "isPlaylistAutoplayEnabled", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "openPlaylistAutoplaySettings", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getCurrentPlaybackMetadata", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "loadPlaylist", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "loadTrack", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "next", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "play", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "pause", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "previous", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "seek", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getPlaybackState", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "destroy", returnType: CAPPluginReturnPromise)
    ]

    private static let playlistURLPrefix = "orpheus://playlist/"
    private static let songURLPrefix = "orpheus://song/"
    private static let autoplaySuffix = "/?autoplay=1"

    private var currentTrackId: String?
    private var preparedURL: URL?
    private var externalOpened = false
    private var lastError: String?

    @objc func initialize(_ call: CAPPluginCall) {
        call.resolve()
    }

    @objc func ensureLoggedIn(_ call: CAPPluginCall) {
        call.resolve()
    }

    @objc func configureSourceReveal(_ call: CAPPluginCall) {
        call.resolve()
    }

    @objc func isFloatingWindowEnabled(_ call: CAPPluginCall) {
        call.resolve(["enabled": false])
    }

    @objc func openFloatingWindowSettings(_ call: CAPPluginCall) {
        call.reject("Floating source window is not available on iOS.",
                    "player_action_unsupported")
    }

    @objc func isPlaybackMonitorEnabled(_ call: CAPPluginCall) {
        call.resolve(["enabled": false])
    }

    @objc func openPlaybackMonitorSettings(_ call: CAPPluginCall) {
        call.reject("NetEase playback monitoring is not available on iOS.",
                    "player_action_unsupported")
    }

    @objc func isPlaylistAutoplayEnabled(_ call: CAPPluginCall) {
        call.resolve(["enabled": false])
    }

    @objc func openPlaylistAutoplaySettings(_ call: CAPPluginCall) {
        call.reject("NetEase playlist autoplay is not available on iOS.",
                    "player_action_unsupported")
    }

    @objc func getCurrentPlaybackMetadata(_ call: CAPPluginCall) {
        call.resolve([
            "status": "unsupported",
            "updated_at_ms": Int(Date().timeIntervalSince1970 * 1000)
        ])
    }

    @objc func loadTrack(_ call: CAPPluginCall) {
        guard let songId = call.getString("netease_song_id"), !songId.isEmpty else {
            call.reject("netease_song_id is required.", "invalid_track_id")
            return
        }
        currentTrackId = songId
        prepareURL(Self.songURLPrefix + songId + Self.autoplaySuffix)
        call.resolve()
    }

    @objc func loadPlaylist(_ call: CAPPluginCall) {
        guard let playlistId = call.getString("netease_playlist_id"), !playlistId.isEmpty else {
            call.reject("netease_playlist_id is required.", "invalid_playlist_id")
            return
        }
        currentTrackId = nil
        prepareURL(Self.playlistURLPrefix + playlistId + Self.autoplaySuffix)
        call.resolve()
    }

    @objc func play(_ call: CAPPluginCall) {
        guard let preparedURL else {
            call.reject("No NetEase playlist or song is prepared.", "player_not_initialized")
            return
        }
        DispatchQueue.main.async {
            UIApplication.shared.open(preparedURL) { opened in
                if opened {
                    self.externalOpened = true
                    self.lastError = nil
                    call.resolve()
                    return
                }
                self.lastError = "netease_app_open_failed"
                call.reject("NetEase app could not be opened.", "netease_app_open_failed")
            }
        }
    }

    @objc func pause(_ call: CAPPluginCall) {
        rejectExternalControl(call)
    }

    @objc func next(_ call: CAPPluginCall) {
        rejectExternalControl(call)
    }

    @objc func previous(_ call: CAPPluginCall) {
        rejectExternalControl(call)
    }

    @objc func seek(_ call: CAPPluginCall) {
        call.reject("Seek is not available when playback is handled by NetEase app.",
                    "player_action_unsupported")
    }

    @objc func getPlaybackState(_ call: CAPPluginCall) {
        if externalOpened {
            call.resolve(playbackState("playing"))
            return
        }
        if preparedURL != nil {
            call.resolve(playbackState("paused"))
            return
        }
        call.resolve(playbackState("idle"))
    }

    @objc func destroy(_ call: CAPPluginCall) {
        currentTrackId = nil
        preparedURL = nil
        externalOpened = false
        lastError = nil
        call.resolve()
    }

    private func prepareURL(_ value: String) {
        preparedURL = URL(string: value)
        externalOpened = false
        lastError = nil
    }

    private func rejectExternalControl(_ call: CAPPluginCall) {
        call.reject("Transport controls are not available in external NetEase app mode.",
                    "player_action_unsupported")
    }

    private func playbackState(_ state: String) -> [String: Any] {
        return [
            "state": state,
            "currentTimeMs": 0,
            "durationMs": 0,
            "currentTrackId": currentTrackId as Any,
            "lastError": lastError as Any
        ]
    }
}
